### PR TITLE
docs: add Developer document for Performance Profiling in zot

### PR DESCRIPTION
**Pull request by mbshields**
**What type of PR is this?**

documentation

**Which issue does this PR fix**:

[Feat]: Include a profiling article in the docs
[127](https://github.com/project-zot/project-zot.github.io/issues/127)

**What does this PR do / Why do we need it**:

explains to developers how to make use of the performance profiling feature in zot

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/developer-guide/pprofiling.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+| mutex | Stack traces of holders of contended mutexes. |
+| profile | CPU usage profile. You can specify the duration in the `seconds` URL query parameter. After receiving the profile file, use the `go tool pprof` command to investigate the profile. |
+| threadcreate | Stack traces that led to the creation of new OS threads. |
+| trace | A trace of execution of the current program. You can specify the duration in the `seconds` URL query parameter. After you get the trace file, use the `go tool trace` command to investigate the trace. |
+
+To return an HTML-format profile list along with a count of currently available records for each profile, use the following API command: 
+
+    /v2/_zot/pprof/
+
+> :pencil2: You must be authenticated to access `/v2/` API commands.
```
Comment (andaaron) [inline docs/developer-guide/pprofiling.md:27]:
```suggestion
> :pencil2: If you have authentication enabled, only an admin user can access this API.
```

File: docs/developer-guide/pprofiling.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
+# Performance Profiling in zot
+
+> :point_right: Use zot's built-in profiling tools to collect and analyze runtime performance.
+
+The profiling capabilities within zot allow a zot administrator to collect and export a range of diagnostic performance data such as CPU intensive function calls, memory allocations, and execution traces. The collected data can then be analyzed using go tools and a variety of available visualization tools.
```
Comment (rchincha) [inline docs/developer-guide/pprofiling.md:5]: would add a link from "administrator" to authN/authZ content

File: docs/developer-guide/pprofiling.md (hunk 1 above)
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:27]: Done: "If authentication is enabled, only an admin user can access this API."

File: docs/developer-guide/pprofiling.md (hunk 2 above)
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:5]: Done: "...allow a zot [administrator](../articles/authn-authz.md) to collect..."

File: docs/developer-guide/pprofiling.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+The pprof package can generate graphic representations of profile data in many formats. This example generates a GIF file representing the CPU usage in the "cpu10.prof" file.
+
+    $ go tool pprof -gif cpu10.prof
+    Generating report in profile001.gif
+
+![profile001.gif](../assets/images/profile001.gif){width="300"}
+
+!!! note "Documentation comments"
+
+    Is this section helpful?  Need a more interesting illustration (maybe Flame?)
```
[the remark this answers is not in the document]
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:96]: Need a more typical example, if possible.

**Comment by rchincha** [comment]
Put profiling document under "Articles" and cross-ref from admin guide (since admin creds/authZ) and developer guide.

File: docs/developer-guide/pprofiling.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    /v2/_zot/pprof/
+
+> :pencil2: If authentication is enabled, only an admin user can access this API.
+
+## How do I export profile data?
+
+To collect and export any available profile, use the following API command format:
+
+    /v2/_zot/pprof/<profile-name>[?<query-parameters>]
```
Comment (rchincha) [inline docs/developer-guide/pprofiling.md:33]: <profile-type>

File: docs/developer-guide/pprofiling.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+| profile | CPU usage profile. You can specify the duration in the `seconds` URL query parameter. After receiving the profile file, use the `go tool pprof` command to investigate the profile. |
+| threadcreate | Stack traces that led to the creation of new OS threads. |
+| trace | A trace of execution of the current program. You can specify the duration in the `seconds` URL query parameter. After you get the trace file, use the `go tool trace` command to investigate the trace. |
+
+To return an HTML-format profile list along with a count of currently available records for each profile, use the following API command: 
+
+    /v2/_zot/pprof/
+
+> :pencil2: If authentication is enabled, only an admin user can access this API.
+
```
Comment (rchincha) [inline docs/developer-guide/pprofiling.md:28]: ## Supported Profile Types

`zot` incorporates [golang`s pprof package](https://pkg.go.dev/runtime/pprof) and therefore the following profile types are supported.

profile
heap
...

File: docs/developer-guide/pprofiling.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+## How do I export profile data?
+
+To collect and export any available profile, use the following API command format:
+
+    /v2/_zot/pprof/<profile-name>[?<query-parameters>]
+
+The following example shows an API request for the CPU usage profile named `profile` using a collection window of 10 seconds:
```
Comment (rchincha) [inline docs/developer-guide/pprofiling.md:35]: ... assuming `zot` is running at localhost:8080

Comment (rchincha) [reply docs/developer-guide/pprofiling.md:35]: Or ...
unless specified, the examples and code snippets in this article assume zot running at localhost:8080

File: docs/developer-guide/pprofiling.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+    $ curl -s -v http://localhost:8080/v2/_zot/pprof/profile?seconds=10 > cpu10.prof
```
Comment (rchincha) [inline docs/developer-guide/pprofiling.md:37]: Drop the -v and all the curl logs below.

Comment (rchincha) [reply docs/developer-guide/pprofiling.md:37]: " > cpu.prof"

File: docs/developer-guide/pprofiling.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    < Content-Disposition: attachment; filename="profile"
+    < Content-Type: application/octet-stream
+    < X-Content-Type-Options: nosniff
+    < Date: Tue, 26 Sep 2023 17:01:28 GMT
+    < Content-Length: 107
+    < 
+    { [107 bytes data]
+    * Connection #0 to host localhost left intact
+
+This command creates an output data file named "cpu10.prof".
```
Comment (rchincha) [inline docs/developer-guide/pprofiling.md:55]: "cpu.prof"

File: docs/developer-guide/pprofiling.md [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+A Flame Graph can be very useful for analyzing CPU usage:
+
+![flame.svg](../assets/images/flame.svg){width="300"}
+
+
+### Generating a graphic image
+
+The pprof package can generate graphic representations of profile data in many formats. This example generates a GIF file representing the CPU usage in the "cpu10.prof" file.
+
+    $ go tool pprof -gif cpu10.prof
```
Comment (rchincha) [inline docs/developer-guide/pprofiling.md:89]: "cpu.prof"

File: docs/developer-guide/pprofiling.md [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+## Analyzing the trace profile using `go tool trace`
+
+!!! note "Documentation comments"
+
+    Is this section helpful? 
+
+You can collect trace data with the `trace` profile, as in this example:
+
+      $ curl -s -v http://localhost:8080/v2/_zot/pprof/trace?seconds=10 > trace10.prof
```
Comment (rchincha) [inline docs/developer-guide/pprofiling.md:133]: "trace.prof"

Comment (rchincha) [reply docs/developer-guide/pprofiling.md:133]: Also standardize "seconds=30"

File: mkdocs.yml [hunk 11]
```diff
@@ -110,6 +110,7 @@ nav:
   - Developer Guide: 
     - Onboarding: developer-guide/onboarding.md
     - Extensions: developer-guide/extensions-dev.md
+    - Performance Profiling: developer-guide/pprofiling.md
```
Comment (rchincha) [inline mkdocs.yml:113]: Move this under Articles and xref.

File: docs/developer-guide/pprofiling.md [hunk 12]
```diff
@@ -0,0 +1,146 @@
+# Performance Profiling in zot
```
Comment (rchincha) [inline docs/developer-guide/pprofiling.md:1]: Instead of assets/profile001.gif 
Convert this into jpg or png

Rename this file and profiling-001.png

Comment (rchincha) [reply docs/developer-guide/pprofiling.md:1]: Rename assets/images/flame.svg and profiling-flame.svg

File: docs/developer-guide/pprofiling.md (hunk 4 above)
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:33]: Changed.

File: docs/developer-guide/pprofiling.md (hunk 6 above)
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:35]: Added note in the top section:  "All examples in this article assume that the zot registry is running at `localhost:8080`."

File: docs/developer-guide/pprofiling.md (hunk 7 above)
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:37]: Changes made.

File: docs/developer-guide/pprofiling.md (hunk 8 above)
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:55]: Done

File: docs/developer-guide/pprofiling.md (hunk 9 above)
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:89]: Done

File: docs/developer-guide/pprofiling.md (hunk 10 above)
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:133]: Done

File: mkdocs.yml (hunk 11 above)
Comment (mbshields) [reply mkdocs.yml:113]: Done.

File: docs/developer-guide/pprofiling.md (hunk 12 above)
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:1]: Done

File: docs/developer-guide/pprofiling.md (hunk 4 above)
Comment (mbshields) [reply docs/developer-guide/pprofiling.md:33]: Moved to "Articles" and added cross-ref links from other guides.

File: docs/general/whats-new.md [hunk 13]
```diff
@@ -52,6 +52,14 @@
 
 - You can adjust the background scheduler based on your deployment requirements for tasks that are handled in the background, such as garbage collection.  See [Configuring zot](../admin-guide/admin-configuration.md).
 
+### Performance profiling for troubleshooting
```
Comment (rchincha) [inline docs/general/whats-new.md:55]: same thing copied twice here?

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**